File: Stripe/STPConfirmCardOptions.m
//
//  STPConfirmCardOptions.m
//  Stripe
//
//  Created by Cameron Sabol on 1/10/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

#import "STPConfirmCardOptions.h"

NS_ASSUME_NONNULL_BEGIN

@implementation STPConfirmCardOptions

@synthesize additionalAPIParameters;

- (NSString *)description {
    NSMutableArray *props = [@[
                               // Object
                               [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],

                               [NSString stringWithFormat:@"cvc = %@", self.cvc],
                               [NSString stringWithFormat:@"network = %@", self.network],
                               ] mutableCopy];


    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPFormEncodable

+ (nonnull NSDictionary *)propertyNamesToFormFieldNamesMapping {
    return @{
        NSStringFromSelector(@selector(cvc)): @"cvc",
        NSStringFromSelector(@selector(network)): @"network",
    };
}

+ (nullable NSString *)rootObjectName {
    return @"card";
}

@end

NS_ASSUME_NONNULL_END

File: StripePayments/StripePayments/Source/API Bindings/Models/PaymentIntents/STPConfirmCardOptions.swift
//
//  STPConfirmCardOptions.swift
//  StripePayments
//
//  Created by Cameron Sabol on 1/10/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

import Foundation

/// Options to update a Card PaymentMethod during PaymentIntent confirmation.
/// - seealso: https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-payment_method_options-card
public class STPConfirmCardOptions: NSObject {

    /// CVC value with which to update the Card PaymentMethod.
    @objc public var cvc: String?

    /// Selected network to process this PaymentIntent on. Depends on the available networks of the card attached to the PaymentIntent. Can be only set confirm-time.
    @objc public var network: String?

    /// :nodoc:
    @objc public var additionalAPIParameters: [AnyHashable: Any] = [:]

    /// :nodoc:
    @objc public override var description: String {
        let props = [
            // Object
            String(format: "%@: %p", NSStringFromClass(type(of: self)), self),
            "cvc = \(String(describing: cvc))",
            "network = \(String(describing: network))",
        ]

        return "<\(props.joined(separator: "; "))>"
    }

}

// MARK: - STPFormEncodable
extension STPConfirmCardOptions: STPFormEncodable {

    @objc
    public class func propertyNamesToFormFieldNamesMapping() -> [String: String] {
        return [
            NSStringFromSelector(#selector(getter: cvc)): "cvc",
            NSStringFromSelector(#selector(getter: network)): "network",
        ]
    }

    @objc
    public class func rootObjectName() -> String? {
        return "card"
    }
}
